R/tmb-methods.R

```diff
@@ -429,3 +429,114 @@ h_residuals_response <- function(object) {
   assert_class(object, "mmrm_tmb")
   component(object, "y_vector") - unname(fitted(object))
 }
+
+
+#' @describeIn mmrm_tmb_methods Simulates from an mmrm model object
+#' @importFrom stats simulate
+#' @exportS3Method
+#'
+#' @param object an object of class 'mmrm'
+#' @param nsim number of response vectors to simulate. Defaults to 1.
+#' @param seed an object specifying if and how the random number generator should be initialized (‘seeded’).
+#' @param newdata data frame for which to evaluate predictions.
+#' @param ...
+#'
+#' @return A \code{data.frame} of dimension [n, m] where n is the number of observations
+#' fitted in the model, and m is the number \code{nsim} of simulated responses.
+simulate.mmrm_tmb <- function(object, nsim = 1,
+                              newdata = NULL,
+                              method = c("conditional", "marginal"),
+                              ...){
+
+  method <- match.arg(method)
+
+  if(is.null(newdata)){
+    # use data in fit
+    newdata <- object$data
+  }
+
+  # process the new data -- make sure new data has the same levels as original data
+  full_frame <- model.frame(
+    object,
+    data = newdata,
+    include = c("subject_var", "visit_var", "group_var", "response_var"),
+    na.action = "na.pass"
+  )
+  tmb_data <- h_mmrm_tmb_data(
+    object$formula_parts, full_frame,
+    weights = rep(1, nrow(full_frame)),
+    reml = TRUE,
+    singular = "keep",
+    drop_visit_levels = FALSE,
+    allow_na_response = TRUE,
+    drop_levels = FALSE
+  )
+
+  if(method == "conditional"){
+    # order data.frame
+    # TODO remove -- conform with predict() method approach
+    newdata <- dplyr::arrange(newdata, !!object$formula_parts$subject_var, !!object$formula_parts$visit_var)
+
+    # get prediction and prediction variance
+    mu <- h_get_prediction(tmb_data, object$theta_est, object$beta_est, object$beta_vcov)
+
+    # build data.frame of prediction vectors
+    ret <- mu$prediction[,1]
+    ret <- data.frame(replicate(nsim, ret))
+
+    # get dataframe values
+    n_subjects <- tmb_data$n_subjects
+    subject_n_visits <- tmb_data$subject_n_visits
+    subject_zero_inds <- tmb_data$subject_zero_inds
+    visit_zero_inds <- tmb_data$visits_zero_inds
+
+      for(i in 1:n_subjects){
```

Comment:
```suggestion
      for(i in seq_len(n_subjects)) {
```

seq_len is usually robust than 1:n_subjcts